File: api/grade.py
from flask import Flask, request, jsonify
from dashscope import MultiModalConversation, Generation
import dashscope
import os
import re

# 从环境变量读取 API Key
dashscope.api_key = os.getenv("DASHSCOPE_API_KEY")

def is_multiple_choice(question: str) -> bool:
    patterns = [r'\b[A-D]\.\s', r'\([A-D]\)', r'选项[：:]?\s*[A-D]', r'四个选项']
    return any(re.search(p, question, re.IGNORECASE) for p in patterns)

def handler(request):
    if request.method != 'POST':
        return jsonify({"error": "Only POST allowed"}), 405

    data = request.get_json()
    question = data.get('question', '').strip()
    user_answer = data.get('user_answer', '').strip()
    image_base64 = data.get('image', '')

    if not question or not user_answer:
        return jsonify({"error": "题目和答案不能为空"}), 400

    is_mc = is_multiple_choice(question)
    system_prompt = (
        "你是一位严谨的学科教师，请对以下题目进行批改。\n"
        "请严格按以下格式输出：\n"
        "【判断】正确/错误\n"
        f"【{'正确答案' if is_mc else '参考答案'}】...\n"
        "【解析】...\n"
        "【考点】...\n"
        "【错因】..."
    )
    user_prompt = f"题目：{question}\n我的答案：{user_answer}"

    try:
        if image_base64:
            messages = [
                {'role': 'system', 'content': system_prompt},
                {'role': 'user', 'content': [{'image': image_base64}, {'text': user_prompt}]}
            ]
            response = MultiModalConversation.call(model='qwen-vl-plus', messages=messages)
        else:
            response = Generation.call(
                model='qwen-plus',
                messages=[{'role': 'system', 'content': system_prompt}, {'role': 'user', 'content': user_prompt}]
            )

        if response.status_code == 200:
            ai_text = response.output.choices[0].message.content
            return jsonify({"ai_feedback": ai_text})
        else:
            return jsonify({"error": f"AI 错误: {response.code}"}), 500
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Vercel 入口
def grade(req):
    from vercel_wsgi import serve
    app = Flask(__name__)
    CORS(app)  # 注意：Vercel 中需处理 CORS
    @app.route('/api/grade', methods=['POST'])
    def route():
        return handler(req)
    return serve(app, req)

File: api/grade.js
// api/grade.js
import { NextResponse } from 'next/server';
import { Qwen } from '@alicloud/dashscope-sdk';

const client = new Qwen({
  apiKey: process.env.DASHSCOPE_API_KEY,
});

function isMultipleChoice(question) {
  const patterns = [
    /\b[A-D]\.\s/i,
    /\([A-D]\)/i,
    /选项[：:]?\s*[A-D]/i,
    /四个选项/i
  ];
  return patterns.some(p => p.test(question));
}

export async function POST(request) {
  try {
    const data = await request.json();
    const { question, user_answer: userAnswer, image } = data;

    if (!question || !userAnswer) {
      return NextResponse.json({ error: "题目和答案不能为空" }, { status: 400 });
    }

    const isMc = isMultipleChoice(question);
    const systemPrompt = 
      ` "你是一名严谨的中学教师，请根据题目类型采用以下规则批改：\n\n"
        
        "📌 如果题目是【选择题】（明显包含 A、B、C、D 等选项）：\n"
        "- 【正确答案】仅写出标准选项字母（如：C）；\n"
        "- 【解析】先直接说明“正确选项是 X，因为……”，然后逐条分析其他每个选项为何错误。\n\n"
        
        "📌 如果题目是【非选择题】（如填空题、计算题、解答题等）：\n"
        "- 【正确答案】写出完整的标准答案；\n"
        "- 【解析】给出完整、规范的正确解题过程，步骤清晰,简单明了。\n\n"
        
        "此外，所有题目都必须包含：\n"
        "- 【考点】指出考查的具体知识点；\n"
        "- 【错因】结合学生答案，判断是“粗心”还是“概念不清”。\n\n"
        
        "请严格按以下五点格式作答，每点独立成段：\n"
        "1. 【判断】\n"
        "2. 【正确答案】\n"
        "3. 【解析】\n"
        "4. 【考点】\n"
        "5. 【错因】\n\n"

    const userPrompt = `题目：${question}\n我的答案：${userAnswer}`;

    let response;
    if (image) {
      // 图像+文本（暂不支持，简化为文本）
      response = await client.chat.completions.create({
        model: 'qwen-plus',
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt }
        ]
      });
    } else {
      response = await client.chat.completions.create({
        model: 'qwen-plus',
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt }
        ]
      });
    }

    const aiFeedback = response.choices[0].message.content;
    return NextResponse.json({ ai_feedback: aiFeedback });

  } catch (error) {
    console.error('AI Error:', error);
    return NextResponse.json({ error: error.message || '批改失败' }, { status: 500 });
  }
}
